use a HANDSHAKE_DONE frame to drive the handshake to confirmation (#3145)
quicwg/base-drafts: draft-ietf-quic-tls.md

### Handshake Confirmed {#handshake-confirmed}

In this document, the TLS handshake is considered confirmed at the server when
the handshake completes. At the client, the handshake is considered confirmed

> when the HANDSHAKE_DONE frame is received.

Reviewer: I don't think that we should add an "OR" here.  If the server acts badly, then the handshake won't ever be confirmed and that's OK.

> when the HANDSHAKE_DONE frame is received.

Reviewer 2: If we're going to add a new signal, I'd like to enforce that.  And if the rule is that the server must send HANDSHAKE_DONE in a short period of time, then it's reasonable for the client to assume any ACK in 1-RTT prior to receiving the HANDSHAKE_DONE frame is a result of the packet containing HANDSHAKE_DONE being lost.

I don't see a downside of this addition, given it's quite simple.

> when the HANDSHAKE_DONE frame is received.

Reviewer 3: Overall, I quite like this HANDSHAKE_DONE proposal. It seems obviously correct which, given our previous iterations is rather appealing :)

I agree with Ian that allowing the client to drop handshake keys when it receives an ACK of a 1-RTT packet sounds like a great optimization which helps ensure the keys get dropped quickly.

> when the HANDSHAKE_DONE frame is received.

Reviewer 4: I guess the only danger is that putting in this backstop makes it semi-OK for servers to not bother to send HANDSHAKE_DONE as an optimization, and then all the client data goes in 0RTT packets and we end up in the deadlock.

I would rather noncompliance result in holding keys, rather than good connections timing out.

> when the HANDSHAKE_DONE frame is received.

Reviewer 2: The server could also not send anything in Handshake and force the client to continue using 0RTT packets, which seems equivalent to what you're describing?  If so, then this is nothing new.

Maybe we need more motivating text about why you really need to send HANDSHAKE_DONE, but I think this design is generally quite efficient.

If we REALLY want to ensure the client gets the signal, we can add a key phase flip between 0.5RTT and 1RTT as well, but then do we change keys as well?

> when the HANDSHAKE_DONE frame is received.

Reviewer 4: @ianswett That's definitely not what I was talking about. But I see now that my example is not right.

> when the HANDSHAKE_DONE frame is received.

Reviewer 5: nit: could we say "when *a* HANDSHAKE_DONE frame is received."

> 

Reviewer 6: Is it clear in which PN space the HANDSHAKE_DONE frame is sent in? To me it seems it could be in either 1-RTT or HS PN. But going forward - since HS keys must be dropped and done must be transmitted, it then follows it must be 1-RTT.

Reviewer 7: Handshake is done at the server upon receipt of the ClientFinished.  Once it's done, the frame is transmitted by the server, and the server is using 1-RTT keys to send.  Therefore, it has to be 1-RTT.

However, I can no longer find the section specifying which frame types are allowed in which encryption level.  This should probably be listed there, and if we've lost that text, we might have bigger problems.

Author: https://quicwg.org/base-drafts/draft-ietf-quic-tls.html#rfc.section.4 says
> All other frame types MUST only be sent in the 0-RTT and 1-RTT levels.

And since the frame can't be sent by the client, this means that the frame can only be sent in 1-RTT packets.

### Handshake Confirmed {#handshake-confirmed}

In this document, the TLS handshake is considered confirmed at the server when
the handshake completes.  At the client, the handshake is considered confirmed
when a HANDSHAKE_DONE frame is received.

A client MAY consider the handshake to be confirmed when it receives an
acknowledgement for a 1-RTT packet.  This can be implemented by recording the

> lowest packet number sent with 1-RTT keys, and the highest value of the Largest

Reviewer 2: ```suggestion
lowest packet number sent with 1-RTT keys, and comparing to the Largest
```

> Acknowledged field in any received 1-RTT ACK frame: once the latter is higher

Reviewer 7: ```suggestion
Acknowledged field in any received 1-RTT ACK frame: once the latter is greater
```

> than or equal to the former, the handshake can be confirmed.

Reviewer 7: ```suggestion
than or equal to the former, the handshake is confirmed.
```
I see the discussion below; I think it's better as "is," since you're describing how it would be implemented.  However, that's certainly not a hard blocker.

Reviewer 8: I agree. I think this should be "is".

### Discarding Handshake Keys

> An endpoint MUST discard its handshake keys when the TLS handshake is confirmed

Reviewer: A "MAY" suffices here, does it not?  Or are you concerned about having to think about Handshake keys after migration?

It could be enough to say that Handshake packets MUST NOT be sent on any path other than the one on which the connection was established.

Author: There are a few things that could go wrong here, if keys aren't dropped. One is sending Handshake packets after a migration, another one is that a peer could retransmit Handshake packets after the handshake is confirmed, thereby consuming the complete congestion window.
Is there any reason to not make it mandatory?

Reviewer 8: I am concerned about Handshake packets after migration, but I think @martinthomson's text seems more appropriate as a requirement. How about saying instead that Handshake packets MUST NOT be sent after the TLS handshake is confirmed?

Reviewer 9: +1 to "MUST NOT be sent". At the same time, I do not think we need to specify when the key is discarded. Maybe "can discard" is fine?

Author: I'm not opposed to making this change, but I noticed that in the previous paragraph, we specify the dropping of Initial keys:

> Thus, a client MUST discard Initial keys when it first sends a Handshake packet and a server MUST discard Initial keys when it first successfully processes a Handshake packet.  Endpoints MUST NOT send Initial packets after this point.

Why do we use a MUST there, but not for Handshake keys? To prevent packet injections, it would be sufficient to say "MUST NOT process Initial packets". Should we make it consistent for both encryption levels?

Reviewer: This isn't about consistency, it is about necessity.

Reviewer 6: By having MUST dropped (or not further processes which to me is the same), there cannot be replayed packet attacks from handshake or injected packets by partially trusted co-processing logic to setup connections. I'm not sure these are valid concerns, but I don't see any benefits in keeping the handshake either. Notably, coalesced packets with a handshake packet can immediately be dropped if MUST discard is true.

Vague memory: there is a situation where a peer might want to send on invalid packet with a valid CID in order for reflection to work with ECN / RESET or something. I just don't recall what it is, only that I'm the only one disliking this "hack". However, if this "hack" relies on handshake keys being available, it won't work when they are not available, obviously.

Reviewer 2: I'd vote for SHOULD, because I tend to think dropping them immediately is the best option, but I'm not convinced anything terrible happens if a peer delays dropping them a bit.

Reviewer: > @martinthomson I don't understand your comment? What's the difference between dropping Initial keys and not processing them any more?

I was merely pointing out that the Initial keys transition is different in that continuing to handle Initial packets leaves endpoints exposed to attack longer than is ideal.  An attacker with access to Handshake keys most likely can break the remainder of the connection.

(This isn't completely true, because an attacker might only gain leaf keys and not the secrets from the core of the TLS key schedule, but the point was that access to Handshake keys signifies a much more significant breach.)

({{handshake-confirmed}}).  The server MUST send a HANDSHAKE_DONE frame when the
handshake completes.  Most application protocols will send data after the
handshake.  A server might want to minimize the number of packets sent by

> bundling the HANDSHAKE_DONE frame with other data.  It MAY delay sending of the

Reviewer: I think that you can drop the bundling sentence.

> HANDSHAKE_DONE frame by no more than one round-trip.

Reviewer: This delay sentence is somewhat in conflict with the preceding requirement, so maybe instead say "MUST send within 1 round-trip time of handshake completion."

Author: Done.

### Discarding Handshake Keys

An endpoint MUST discard its handshake keys when the TLS handshake is confirmed
({{handshake-confirmed}}).  The server MUST send a HANDSHAKE_DONE frame within

> one round-trip time of handshake completion.

Reviewer 9: I think "within one round-trip time" is too long. As this happens before both endpoints have good estimation of RTT, it could well be the case that the server's estimated RTT be much greater than the PTO of the client. In such case, it would seem to the client as if the CF had been lost.

I think we should say "SHOULD immediately send", with possibly acknowledging the fact that the server "can bundle the HANDSHAKE_DONE frame with 0.5 RTT data."

Reviewer 2: I tend to agree with @kazuho point.  How about "The server MUST send HANDSHAKE_DONE with the first outgoing 1-RTT packet or within one round-trip time of handshake completion if no 1-RTT packets are sent."

@kazuho If the server is sending HANDSHAKE_DONE, isn't it no longer 0.5RTT data?

Reviewer 4: I agree with @ianswett. The handshake isn't complete when we have 0.5RTT data. In fact, if the frame goes out then, we can easily have the client discard the handshake keys before getting ACK of its finished.

Reviewer 9: @ianswett 
> If the server is sending HANDSHAKE_DONE, isn't it no longer 0.5RTT data?

Yes. This is 1 RTT, *not* 0.5 RTT.

Regarding when HANDSHAKE_DONE should be sent, my view is that it HANDSHAKE_DONE is essentially an ACK for the Handshake packet. Based on that view, I do not see a reason why it should be sent at a different timing than other acks for Handshake packets (i.e. send immediately, though an endpoint might wait for a short amount of time to bundle ACK with some other data).

In practice, there is no reason to delay the emission of HANDSHAKE_DONE as much as 1 RTT in HTTP/3. The server should be sending HANDSHAKE_DONE at an early moment for letting the client know that the handshake is done before the client retransmits CF, doing so has a higher probability of having CWND increased at an earlier point.
